### PR TITLE
feat: configurable expiration policy on auto-created subscriptions (#2876) (@jmitash)

### PR DESCRIPTION
**Pull request by burkedavison**
Original PR #2894

Author: @jmitash

* Add properties for Binder expiration policy
* Set Expiration Policy when creating Subscription
* Treat non-positive durations as "never expire"

### CODE REVIEW

File: spring-cloud-gcp-pubsub-stream-binder/src/main/java/com/google/cloud/spring/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java [hunk 1]
```diff
@@ -101,7 +100,7 @@ public ConsumerDestination provisionConsumerDestination(
         subscriptionName = "anonymous." + topicShortName + "." + UUID.randomUUID();
         this.anonymousGroupSubscriptionNames.add(subscriptionName);
       }
-      ensureSubscriptionExists(subscriptionName, topicName, deadLetterPolicy, autoCreate);
+      ensureSubscriptionExists(subscriptionName, topicName, properties.getExtension(), autoCreate);
```
Comment (meltsufin) [inline spring-cloud-gcp-pubsub-stream-binder/src/main/java/com/google/cloud/spring/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java:103]: `autoCreate` can be retrieved from `properties.getExtension()`.

Comment (burkedavison) [reply spring-cloud-gcp-pubsub-stream-binder/src/main/java/com/google/cloud/spring/stream/binder/pubsub/provisioning/PubSubChannelProvisioner.java]: Done

File: docs/src/main/asciidoc/pubsub.adoc [hunk 2]
```diff
@@ -426,6 +426,29 @@ public Subscription newSubscription() {
 
 Dead letter topics are no different from any other topic, though some https://cloud.google.com/pubsub/docs/dead-letter-topics#granting_forwarding_permissions[additional permissions] are necessary to ensure the Cloud Pub/Sub service can successfully `ack` the original message and re-`publish` on the dead letter topic.
 
+===== Expiration Policy
```
[the remark this answers is not in the document]
Comment (burkedavison) [reply docs/src/main/asciidoc/pubsub.adoc:429]: Done

File: docs/src/main/asciidoc/pubsub.adoc [hunk 3]
```diff
@@ -426,6 +426,27 @@ public Subscription newSubscription() {
...
+
+By default, subscriptions expire after 31 days without any subscriber activity or subscription
+property changes. A `null` or unset `ExpirationPolicy` will use the default.
+
+To set the subscription to `never expire`, provide an `ExpirationPolicy` without setting the TTL.
```
Comment (meltsufin) [inline docs/src/main/asciidoc/pubsub.adoc:434]:
```suggestion
To create a subscription that never expires, provide an `ExpirationPolicy` without setting the TTL.
```

**Review by meltsufin — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/pubsub.adoc [hunk 4]
```diff
@@ -426,6 +426,27 @@ public Subscription newSubscription() {
 
 Dead letter topics are no different from any other topic, though some https://cloud.google.com/pubsub/docs/dead-letter-topics#granting_forwarding_permissions[additional permissions] are necessary to ensure the Cloud Pub/Sub service can successfully `ack` the original message and re-`publish` on the dead letter topic.
 
+===== Expiration Policy
+
+By default, subscriptions expire after 31 days without any subscriber activity or subscription
+property changes. A `null` or unset `ExpirationPolicy` will use the default.
```
[the remark this answers is not in the document]
Comment (burkedavison) [reply docs/src/main/asciidoc/pubsub.adoc]: Done

**Review by meltsufin — APPROVED** [review]

**Pull request merged**